src/main/groovy/org/xbib/gradle/plugin/JFlexPlugin.groovy

```diff
@@ -52,36 +53,66 @@ class JFlexPlugin implements Plugin<Project> {
         }
     }
 
-    private static void createJflexExtension(Project project) {
+    private static JFlexExtension createJflexExtension(Project project) {
         project.extensions.create ('jflex', JFlexExtension)
     }
 
-    private static void addJFlexTasks(Project project) {
-        project.sourceSets.all { SourceSet sourceSet ->
-            addJFlexTaskForSourceSet(project, sourceSet)
+    private static void addJFlexTasks(Project project, JFlexExtension jFlexExtension) {
+        project.extensions.getByType(SourceSetContainer).all { SourceSet sourceSet ->
+            addJFlexTaskForSourceSet(project, sourceSet, jFlexExtension)
         }
     }
 
-    private static void addJFlexTaskForSourceSet(Project project, SourceSet sourceSet) {
+    private static void addJFlexTaskForSourceSet(Project project, SourceSet sourceSet, JFlexExtension jFlexExtension) {
         String taskName = sourceSet.getTaskName('generate', 'jflex')
         SourceDirectorySet sourceDirectorySet = sourceSet.extensions.getByName('jflex') as SourceDirectorySet
-        File targetFile = project.file("${project.buildDir}/generated/sources/${sourceSet.name}")
-        if (sourceDirectorySet.asList()) {
-            TaskProvider<JFlexTask> taskProvider = project.tasks.register(taskName, JFlexTask) {
-                group = 'jflex'
-                description = 'Generates code from JFlex files in ' + sourceSet.name
-                source = sourceDirectorySet.asList()
-                target = targetFile
-                theSourceSet = sourceSet
-            }
-            logger.info "created ${taskName} for sources ${sourceDirectorySet.asList()} and target ${targetFile}"
-            project.tasks.named(sourceSet.compileJavaTaskName).configure({
-                dependsOn taskProvider
-            })
-            if (sourceSet.java && sourceSet.java.srcDirs) {
-                sourceSet.java.srcDirs += targetFile
+        Provider<Directory> defaultTargetFile = project.layout.buildDirectory.dir("generated/sources/${sourceSet.name}")
+
+        Provider<File> targetFile = jFlexExtension.writeIntoJavaSrc.map {
```

Comment:
handles resolution of the target file in a `Provider` rather that is passed to the task rather than having the task provided the whole source set, conditional on the value of `writeIntoJavaSrc` at task execution time and whether or not there are existing java source dirs to write into